pkg/apis/kubeone/validation/validation.go

```diff
@@ -177,6 +179,27 @@ func ValidateVersionConfig(version kubeone.VersionConfig, fldPath *field.Path) f
 	return allErrs
 }
 
+func ValidateContainerRuntimeConfig(cr kubeone.ContainerRuntimeConfig, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allCRs := []interface{}{
+		cr.Docker,
+		cr.Containerd,
+	}
+
+	var found bool
+	for _, x := range allCRs {
+		if !reflect.ValueOf(x).IsNil() {
+			if found {
+				allErrs = append(allErrs, field.Invalid(fldPath, x, "only 1 container runtime can be activated"))
+			}
+			found = true
+		}
+	}
+
+	return allErrs
```

Comment:
As [validation happens after defaulting](https://github.com/kubermatic/kubeone/blob/3835bcac539cd4df037edd0b7505542d622f1fae/pkg/apis/kubeone/config/config.go#L165-L168), we should test is ContainerRuntimeConfig set, i.e. something like:

```
if !found {
    // err...
}
```

Reply:
if validation happens after defaulting, it means that at least `cr.Docker` will be set (not nil), so `if !found` situation is ruled out.

Reply:
While I agree with that, I think validating could be useful to confirm that defaulting works (e.g. can help to catch potential bugs). However, I'm not going to hold on this.